### ISSUE TITLE
LivenessProbe should start after ReadinessProbe Succeeded if ReadinessProbe is specified

### ISSUE THREAD

**Issue by tanapoln** [kubernetes/kubernetes#27114]
Our initial understanding is that liveness probe will start to check after readiness probe was succeeded but it turn out not to be like that.

We are testing with our system that has a long boot time, an approximation of boot time is between 1-3 minutes. We specify readiness probe with same url of liveness probe and specify initial delay of liveness probe to be 30 seconds, we found that the pod is killed by failing of liveness probe while readiness probe still in a failure.

So, why don't we specify initial delay of liveness probe to be more than 3 minutes? Well, it might be a case when readiness probe succeed at first minute and after that the pod will fail to do a job before liveness probe will start. So, it will affect running service.

Another point is why we don't only put readiness probe and leave out liveness probe? When readiness probe fail again it might take out of service as well so, it don't cause any affect for running service but it won't be restarted and we have to do manual restart them.

There are some concerns we thought about, here are list:
- What if readiness probe never succeed while starting a pod, can we specify a maximum number of check will be performed and if readiness probe fail more than that number, we may consider that pod should be killed.
- What if someone want to set readiness probe to put the pod in maintenance mode, the pod might be killed by above point if readiness probe fail more than that number or liveness probe will kill it. Maybe, we add one more state of readiness probe called `Maintenance` so that liveness probe should stop working when enter this state and no number of failing readiness probe should not be considered.

What do you think?

**Comment by lostpebble** [comment]
I see this is still not implemented. This would really be much appreciated and I think its a pretty logical way to think about it:

[`readinessProbe`] Is it ready? **No**
-> Okay lets wait a bit and see [still `readinessProbe`]

_Eventually..._
[`readinessProbe`] Is it ready? **Yes**
-> Okay we can serve traffic and run checks to see if it remains alive [`livenessProbe`]

Having liveness probe kill off the pod before it's ready is frustrating and counter-intuitive. As mentioned, that should be handled completely in the readiness probe - with a set limit on time to ready / check iterations. As it stands now I have to create some rough estimates about how long I should delay my liveness probe - which is not always an easy constant to track.

**Comment by kargakis** [comment]
@kubernetes/sig-node-feature-requests

**Comment by fraenkel** [comment]
The change is easy to implement however what is the correct behavior given what is done today.
If liveness runs after readiness, the delays that are there today would be odd. Like you already stated, it is hard to pick a value but people did.

**Comment by lostpebble** [comment]
Understandably, this will cause unexpected behaviour because of the way people have bent their liveness probes around this at the moment.

I suggest an extra variable on `livenessProbe` perhaps? Something like `waitForReadinessProbe`.

And maybe a warning somewhere that on future versions (however far down the line) liveness probes will wait by default. Or just stick with the extra `waitForReadinessProbe` indefinitely... (it just seems like it should be default behaviour)

**Comment by tobilarscheid** [comment]
I  totally like this idea, my general understanding of a `livenessProbe` was that it is only run on ready containers - I was surprised to find out that this is not the case, when running a container with long boot time for the first time. I vote for making `waitForReadinessProbe` the default, as it makes so much more sense.

**Comment by yujuhong** [comment]
Copying and pasting my own comment from the PR: 
IIUC, liveness was intended to start before readiness in many cases. A process is running doesn't mean it's ready to serve traffic. It won't be restarted (since it passes liveness checks), but will not be listed as endpoints in a service.

Changing the interaction of these two checks will surprise many users.

**Comment by lostpebble** [comment]
@yujuhong I can see that and its a fair understanding. But it doesn't change the fact that there is no easy way to ask your liveness probes to wait until your instances are actually ready- hence causing an early shutdown if you didn't offset it enough. This is not a stable solution at all.

In any case, it seems that the ambiguity of it has caused it to mean different things to different people. I can't think of a time I would consider a pod "alive" but not "ready", so that's why the default behaviour seems strange to me.

I guess the default behaviour should remain the same then. And as suggested, if there is an additional variable such as `waitForReadinessProbe` defined on `livenessProbe`, then the behaviour aligns with what we are expecting here.

Unless there was some other kind of probe that acts as a precursor to both the `readinessProbe` and `livenessProbe` probes. But that would start to feel excessive I think, for such a seemingly easy concept.

**Comment by smarterclayton** [comment]
I don't think it's completely unreasonable for a liveness check to not
start until a readiness check passes (possibly as an option).  But the
edge cases can be complicated (such as on kubelet restart) so it needs
to be sufficiently justified as a usability and "ease of use" feature
to raise above the bar.

**Comment by jrnt30** [comment]
Would it be possible to use a toleration/taint approach for the pod to manage the "state" of this lifecycle in a more persistent way and alleviate the use case of the Kubelet start?  I do understand how this could get a bit tricky, where we may need to also then keep the timestamp of the pods initial Readiness check.  If the pod gets evicted and replaced or restarted, what should happen then, etc. 

In my opinion, as the StatefulSet paradigm continues to increase, this will become an increasingly prevalent issue.  In our case we are attempting to run a few things like Kafka and Elasticsearch which, in an exception case can take dramatically longer to start up (rebuilding indices and rebalancing) than a clean restart would do.  

We are currently evaluating a few options but having the ability to indicate the Liveness check's initial delay should only start after the Readiness passes the first time would take a lot of the guess work out of it for us.

**Comment by gertcuykens** [comment]
hmm what about the scenario where you want this as in if the startup configuration is dynamic waiting for a services that doesn't exist anymore. Then you want to restart the configuration step again to pick up the new changes else it will never boot up

**Comment by jrnt30** [comment]
In that case you could have your readiness probe fail in the first place
and just be polling on that external service being discoverable I think.

**Comment by metametadata** [comment]
Just stumbled upon this problem on rolling out the deployment resource which has a long-ish start-up time. Deployment has both readiness and liveness probes specified. Because the app takes a lot of time to start the failing liveness probe eventually restarts the pod and thus deployment enters the neverending crash/restart loop. This is quite non-intuitive, especially because I use `kubectl rollout status deployment ...` to wait for rollout to finish which seems to depend only on readiness probe.

IMO, at least on deployment rollout, it would be more logical to start probing for *liveness* after deployment's pod is *ready*.

**Comment by zebu** [comment]
I ran into this same issue when first learning K8s. I agree with the previous comments that logically liveness needs to follow readiness.

**Comment by Krzychek** [comment]
How about adding "start before delay if ready" instead of "start when ready" property? so liveness probe would start either after `ready` status or after it has reached its delay.

**Comment by theflyingnerd** [comment]
This problem will probably come up more often with machine learning becoming more popular. These containers can require large downloads and significant loading of data structures into RAM on startup.

It seems there are inherent compromises in each approach. With the proposed approach, you'd never restart machines that hung while they were starting up.

If you want this benefit for a web app with a long start, you will just have to make sure the web app itself is able to start quickly for the liveness check even if there is still state loading though, via threads or external processes. Your liveness check is just your web server returning a simple 200, and your readiness checks that your data structures are loaded into memory.

**Comment by theflyingnerd** [comment]
It makes sense that K8s has treated these separately so that the correspondence between check and its effect is clear.
Using readiness check means you want to automatically change load in response to check.
Using liveness check means you want to automatically restart (depending on policy) in response to check.

I think we should add an extra parameter for the liveness probe `waitForReadinessProbe` to make it wait until readiness has succeeded, but that flag should have a default false so it behaves as before, otherwise there may be cases of containers deadlocking on startup (not passed readiness yet but become dead) that are no longer restarted.

**Comment by MikaelWahlberg** [comment]
No updates on this?  Adding the extra parameter would make a lot of sense. We sometimes do database migrations on pod startup (If something big changed) and that can take an unknown amount of time, so it feels a bit unnecessary to always have the initialDelay very high just to avoid the pod getting killed when doing the migration. If it would wait for readiness instead it would feel much more safe.  I understand that this is not always the case, so this extra parameter would solve the issue for all parties I think.

**Comment by tallclair** [comment]
@MikaelWahlberg - That sounds like something InitContainers might be a better fit for.

Logically this feature request seems backwards to me - how can a container be ready if it's not alive? Also, If the liveness check doesn't run until after readiness has succeeded, then a startup failure would never be detected.

**Comment by MikaelWahlberg** [comment]
@tallclair Ok, you have a point about the InitContainer, will look in to that. 

I guess I can have a separate LivenessProbe for the initContainer, or rather a timeout to solve the startup failure problem.

**Comment by jrnt30** [comment]
There are still some services that require a longer warmup time because they do things in-process.  A few examples I can think of immediately for our stack our Kafka and ElasticSearch, both of which do some integrity checks and data shuffling when the JVM bootstraps the server.  

Both of these have different semantics associated with their startup, but the end result is that we can't reliably use an InitContainer exclusively due to the nature of the underlying process. 

As to the question about Liveness running after Readiness, the Readiness has a timeout as well.  If the pod were to exceed that, that is the signal that a startup error occurred and K8 would take action (reschedule the pod).

**Comment by tallclair** [comment]
> There are still some services that require a longer warmup time because they do things in-process. A few examples I can think of immediately for our stack our Kafka and ElasticSearch, both of which do some integrity checks and data shuffling when the JVM bootstraps the server.

Ack, makes sense. I'm just not sure that liveness waiting for readiness is the right API for it. It almost feels like you want the union of multiple probes. E.g. this pod is alive IF <condition A> OR <condition B>, where one condition is that the pod is still starting up (however you define that) - for example wait for some log line to be printed.

> As to the question about Liveness running after Readiness, the Readiness has a timeout as well. If the pod were to exceed that, that is the signal that a startup error occurred and K8 would take action (reschedule the pod).

I don't follow this. I thought the idea was that liveness would wait until the first readiness check succeeds?

**Comment by jaredallard** [comment]
IMO the idea that we should all change our code to utilize init containers is a bit appalling, and counter intuitive for adoption, as they are a new feature... It's completely reasonable to want a program to only kill itself after it was ready once.

It's also reasonable to want it to NOT be ready but kill it if it takes too long, so I don't see why we can't have another field implementing this feature.

**Comment by andyxning** [comment]
Will work on this later in the next week.

**Comment by andyxning** [comment]
> I think we should add an extra parameter for the liveness probe waitForReadinessProbe to make it wait until readiness has succeeded, but that flag should have a default false so it behaves as before

I am in favor of this suggestion.

**Comment by lostpebble** [comment]
One other thing to consider here is, will the readiness probe ever actually kill off a pod that never actually becomes ready? I assume this was one of the original intentions of the liveness probe.

There still needs to be a kind of "kill switch" if the pod never actually becomes ready.

As @jaredallard mentioned:

> It's also reasonable to want it to NOT be ready but kill it if it takes too long, so I don't see why we can't have another field implementing this feature.

And as @tanapoln mentioned in the first post:

> What if readiness probe never succeed while starting a pod, can we specify a maximum number of check will be performed and if readiness probe fail more than that number, we may consider that pod should be killed.

I think what it boils down to is that the current `livenessProbe` is serving three purposes which should really be split:
* That of killing off a pod that has taken too long to start
* That of checking that a pod is healthy and ready to take traffic now
* And that of checking to make sure our pod is healthy at various intervals

Each of these three things exist in a different time frame and I think its the main cause of confusion here.

To correct this whole thing I would say the liveness probe should only be doing the third thing - the readiness probe should probably be taking care of the first (an amount of retries before declaring unhealthy perhaps) and second one (call the first `livenessProbe` as soon as the first readiness check passes).

Only just saw that @jrnt30 touched on this already:
> As to the question about Liveness running after Readiness, the Readiness has a timeout as well. If the pod were to exceed that, that is the signal that a startup error occurred and K8 would take action (reschedule the pod).

So potentially all we actually do need is the original request of `waitForReadinessProbe` on `livenessProbe` ?

**Comment by tpoindessous** [comment]
Hi 

@andyxning, any update on your work on this issue ?

We have the same problem : we have a lot of Java Web micro services which takes between 1 and 3 minutes to get "Ready".

We use same health URL for livenessProbe and ReadynessProbe and we must set initialDelaySeconds to a larger value, just to be sure that our containers don't be killed when booting.

We wanted to play with failureThreshold in ReadynessProbe to get status quicker. 

But we don't want to touch this value in livenessProbe as it will make detection of failures much longer.

We will very happy to have `waitForReadinessProbe`

Thanks ;)

**Comment by andyxning** [comment]
@tpoindessous Sorry for not working on this recently. Too busy. :(
But i am prepare to work on this in the next week.

**Comment by sfitts** [comment]
We're in the process of migrating our application to K8s and just ran into this exact issue.  Like many others here our application takes a variable amount of time to start and may take longer due to one-time tasks that occur when we migrate from release to release (you could argue these would be better done in an init container, but we're not at the point were we can make things quite that K8s specific).  

The proposed `waitForReadinessProbe` would address our use case nicely.  Any update on the status of that work?

**Comment by knyttl** [comment]
`waitForReadinessProbe` +1

**Comment by matthyx** [comment]
If @andyxning doesn't object, I would like to work on that...
/assign

**Comment by thockin** [comment]
Hi!  I really appreciate the PR, but I'm not sure I buy it.

TL;DR: I think the current API is the right balance of simple and correct.  A more complex API could be possible, but it's not the proposed one and probably not worthwhile.

Details:

If we make this change, there's literally NOTHING that will fix a pod that fails to starts up.  It will hang forever.  That's bad.

These probes are distinct on purpose.  For those pods that take a very long time to start, we have `initialDelaySeconds`.  Set it to a big number.  The only detriment to setting it too big is that you won't restart quickly if you crash during that initial window, but that is literally what you asked for (I don't know whether my pod is still starting up or crashed), and is no worse than the proposed solution.

Put another way, compare the resulting behavior:

Assume today's API.
Set `livenessProbe.initialDelaySeconds = 60` and `readinessProbe.initialDelaySeconds = 10`.

A pod that starts within-bounds:
- t0: not running either probe, ready=false
- t+10s: fail 1st readiness probe
- t+20s: pass readiness probe, traffic can start flowing
- t+21s: crash
- t+30s: fail readiness probe, no new traffic
- t+40s: fail readiness probe, no new traffic
- t+50s: fail readiness probe, no new traffic
- t+60s: fail liveness probe, restart

A pod that fails to start:
- t0: not running either probe, ready=false
- t+10s: fail 1st readiness probe, no traffic
- t+20s: fail readiness probe, no traffic
- t+30s: fail readiness probe, no traffic
- t+40s: fail readiness probe, no traffic
- t+50s: fail readiness probe, no traffic
- t+60s: fail liveness probe, restart

Assume the proposed "waitForReadinessProbe" API.
Set `livenessProbe.waitForReadinessProbe` and `readinessProbe.initialDelaySeconds = 10`.

A pod that starts within-bounds:
- t0: not running either probe, ready=false
- t+10s: fail 1st readiness probe
- t+20s: pass readiness probe, traffic can start flowing
- t+21s: crash
- t+30s: fail liveness probe, restart

A pod that fails to start:
- t0: not running either probe, ready=false
- t+10s: fail 1st readiness probe, no traffic
- t+20s: fail readiness probe, no traffic
- t+30s: fail readiness probe, no traffic
- t+40s: fail readiness probe, no traffic
- t+50s: fail readiness probe, no traffic
- t+60s: fail readiness probe, no traffic
- t+70s: fail readiness probe, no traffic
- t+80s: fail readiness probe, no traffic
.
.
.
forever

So you get slightly better restart behavior during the liveness initialDelay window, at the cost of a much worse failure mode in case of startup failure.

What problem are we trying to solve again?


An idea that I toyed with but never pursued was a `StartupProbe` - all the other probes would wait on it at pod startup.  It could poll on a relatively short period with a long `FailureThreshold`.  Once it is satisfied, the other probes can start.  This produces:

A pod that starts within-bounds:
- t0: not running either probe, ready=false
- t+10s: fail 1st startup probe
- t+15s: pass startup probe
- t+15s: pass 1st liveness probe
- t+15s: fail 1st readiness probe
- t+20s: pass readiness, traffic can start flowing
- t+21s: crash
- t+30s: fail liveness probe, restart

A pod that fails to start:
- t0: not running any probe, ready=false
- t+10s: fail 1st startup probe, no traffic
- t+20s: fail startup probe, no traffic
- t+30s: fail startup probe, no traffic
- t+40s: fail startup probe, no traffic
- t+50s: fail startup probe, no traffic
- t+60s: fail startup probe, restart


This is awkward from a keeping-state POV.  Kubelet needs to know whether it ever saw a startup probe pass, otherwise it could reset the state machine (e.g. if kubelet crashed).

I don't think that this minor semantic improvement is worth the additional complexity or API surface, but I guess I'd be willing to talk it through.  I think that the current API covers almost all cases almost ideally.

**Comment by knyttl** [comment]
You are not describing the situation when the pod take a long and variable time to start. Imagine it starts something between 30-180s. Of course you can set initialDelay to 180s, but the pod might start way sooner and then the liveness probe will be inefficient for two minutes.

**Comment by thockin** [comment]
I did consider that in my examples.  I didn't talk about variability, but I
set the intialDelay to 60.

What is the real impact of the liveness probe having an overly-long initial
delay?  The pod won't have an actual liveness probe for that period of
time, so if the pod hangs or deadlocks, it won't be immediately restarted.
That's JUST at the startup time, and never again thereafter.  Readiness
will still fail, so the impact to end-users should be minimal.  The delay
isn't THAT consequential, since you were willing to wait that long in the
first place.

The question is whether improving that case is really worth the complexity
of more API surface.

**Comment by matthyx** [comment]
> An idea that I toyed with but never pursued was a `StartupProbe` - all the other probes would wait on it at pod startup. It could poll on a relatively short period with a long `FailureThreshold`. Once it is satisfied, the other probes can start.

I kinda like this idea of a `StartupProbe` because it clarifies the purpose of `LivenessProbe` and `ReadinessProbe`.
I guess if we want to go further in this direction we will need a KEP... who's in?

**Comment by ejc3** [comment]
Can’t you simulate StartupProbe with a postStart command that doesn’t return until the container / pod has fully started? The liveness and readiness probes won’t run until that command returns.

**Comment by thockin** [comment]
@ejc3 - That's a awesome idea - I'd have to dig around to verify, but I think you're right.

It would be great if someone would write up a "recipe" for pods that have long startup times, based on this conversation.  Would make a good kubernetes.io blog post....

**Comment by logicalhan** [comment]
@ejc3, but if the liveness probe doesn't run until the postStart command returns, then what would you do in the situation where the container could not achieve proper state during the boot sequence (i.e. the postStart command hangs)? 

I agree with @thockin's earlier statement that the current liveness and readiness concepts probably covers most cases already. That said, if you expect variable start times, this means you need to accommodate more checks failing in succession by lengthening failureThreshold (which is undesirable, since this affects steady state) or you need to delay when you first trigger the liveness check (by increasing the initialDelay). 

> Of course you can set initialDelay to 180s, but the pod might start way sooner and then the liveness probe will be inefficient for two minutes.

This is actually ***not*** a problem in the case where the container comes up quickly (@knyttl). If the container comes up really quickly, then presumably it will have passed the liveness probes. The problem is that when the container fails really fast, the kubelet won't restart the container until that initial window has passed (this is the situation which you want the liveness probe to start sooner). But at best, the startupProbe just allows us to be more efficient about restarting super fast failing containers when we anticipate that our pod **_could_** take a long time to start up. Since we have also have the concept of readiness, when properly used, we can ensure that we send traffic to our container as quickly as we want when it comes up, despite variable in start-up times.

Is there a lot of value in adding the ability to restart a fast failing container?

**Comment by matthyx** [comment]
> Is there a lot of value in adding the ability to restart a fast failing container?

Well, if the application is properly designed for k8s, it should already be able to crash cleanly (and fast) and there is no need to rely on a `LivenessProbe` to kill it.
However, when dealing with legacy applications (J2EE anyone?) I think it would be valuable to have a correct and elegant way to handle that situation using a `StartupProbe`.

**Comment by sfitts** [comment]
> However, when dealing with legacy applications (J2EE anyone?) I think it would be valuable to have a correct and elegant way to handle that situation using a `StartupProbe`.

This, in a nutshell, is the problem.  You can argue all you want about decomposing startup functions into init containers and having applications that start fast and die faster, but... then there is reality.  The system we are migrating to K8s has containers that can take anywhere from 10s to 5m to start, depending on the state of the system, what startup work is needed at a given time, and the hardware that we're running on (we have to support a wide range).  So what's the right setting for the probes?

I'm not saying that the proposed solution is the right one, I'm just saying it would be nice to have a bit more control than we do today.

**Comment by thockin** [comment]
I am trying to wrap my head around the actual impact of the current situation.  I am not convinced that the "my pod crashed in the first 5 minutes and didn't restart" gap has significant enough impact to warrant a new, specialized probe which mutates the pod lifecycle, especially when postStart hooks exist.  If postStart can timeout, it is effectively the same thing, I think?

Our ever-expanding API surface must be reigned in. 
 Convince me that this is THAT important?

**Comment by sfitts** [comment]
> Our ever-expanding API surface must be reigned in.
> Convince me that this is THAT important?

You've clearly made up your mind, so I doubt I can convince you (figured I lend one more voice to try though).  It is import to us, but ultimately that's not what matters here.  IME purity of API is a subjective argument that is hard to win against.

BTW -- what postStart hook would I write?  If it involves adding API surface to a system I don't have access to, how is that going to work?  Not being facetious, would genuinely like to understand the solution you are proposing.  If I can implement the correct semantics myself, without having to change code I can't change and apply that solution generally then it would be equivalent (still be nicer if it can OOB, but that too is subjective).

**Comment by lavalamp** [comment]
Rather than add a second probe or redefine what readiness means, if we wanted to do something here, I'd add a `maxInitialFailureCount int` to the existing probe. With the mechanic that the probe may fail that many times initially but as soon as it succeeds, it will use the ordinary failure count. Kubelet would have to keep a bit of state to implement this.

**Comment by lavalamp** [comment]
Initial delay might be used today for pods where the liveness is flaky while they are starting, so I don't think it's good to change the semantic like that.

**Comment by matthyx** [comment]
So, should I implement a `maxInitialFailureCount int` for the existing probes, would it answer everyone's needs?

**Comment by yra-wtag** [comment]
There is another issue/scenario that we have found out
1. Give the Pods a overestimated liveness delay. Say 120 seconds.
2. Pods is properly started and readiness probe check comes fine and dandy. Great.
3. Now for some strange reason the Pod got deadlocked or for some other reason it is not ready anymore.
4. Now the readiness probe will return `not ready` while liveness probe will return `ready`. The pod will never get restarted wasting resource and money.

So in my humble opinion the liveness probe checking should always depend on the readiness probe. I am seeing that a lot of fellow engineers are asking for the `waitForReadinessProbe` i would ask this to be a time rather than a boolean true false checker. It might work like this for 
`liveness probe`
```
starting check for liveness
    is the pod ready?
         yes!!!
             start checking liveness
         no, sorry!! :( 
              is there a delay set for readiness
                     yes!!!
                          wait for the specified time
                          check  if the pod is ready?
                                   yes!!!
                                            start checking liveness
                                   no, sorry!! :( 
                                            increase fail counter of liveness probe
                      no
                           start checking liveness
```
In this way the old files will not have any issues. While the new variable will increase a lot of issue people are having.

**Comment by lavalamp** [comment]
@yra-wtag:

> Now for some strange reason the Pod got deadlocked or for some other reason it is not ready anymore

This is a bug report about your pod's liveness check, it is failing to express liveness as defined by Kubernetes. Your liveness check should **not** be passing if your pod is deadlocked.

It is possible that your pod, once it goes ready, will never ever go unready--and if it does, it should be restarted--but that is not true for all pods. So, if you want this behavior, build it into your pod's liveness check.

You can use the command based probe to execute arbitrary commands to do the checks. You can make liveness and readiness checks do whatever you need them to! You can even store state (e.g., make an EmptyDir volume) if you want the liveness check to latch and start checking readiness once readiness returns true.

**Comment by lavalamp** [comment]
@matthyx I'd certainly look at an API change PR!

**Comment by yra-wtag** [comment]
@lavalamp 
> This is a bug report about your pod's liveness check, it is failing to express liveness as defined by Kubernetes. Your liveness check should **not** be passing if your pod is deadlocked.

Your argument totally makes sense but then the liveness check is actually checking readiness, don't you think?  Cause any long running process can be considered a Deadlock. as this PR is trying to solve the issue of checking liveness with respect to readiness then there should be a way to define how long a pod should wait for readiness before it considers that as not `recoverable or serviceable` and thus declaring the pod as dead.

**Comment by lavalamp** [comment]
@yra-wtag:

> Your argument totally makes sense but then the liveness check is actually checking readiness, don't you think?

No, I don't agree-- that's not what readiness implies to me at all, as a word (but it's not surprising that I think the word is used correctly since I was probably in the room when we designed the probes in the first place).

For your pod, if it goes unready for a long time, that means something is wrong. But that's definitely not the case for all pods!
* Maybe a dependency is unavailable. To take kube-apiserver and etcd as an example, if etcd goes down, kube-apiserver (should) become unready, but restarting apiserver won't help etcd come up, so marking it unalive would only make things worse. (We actually do this wrong today!)
* Maybe the process takes long-running requests and it's currently serving as many as it has resources to serve; it's not "ready" for new traffic, but restarting it would make it drop the requests it's currently servicing.
* Or maybe we want to drain the process, but it takes a long time: imagine serving an MMO game; you don't want to disconnect existing players but you do need to eventually restart the server shard. So, mark it unready, and wait (maybe a long time) for folks to disconnect.

Kubernetes takes the position that "ready to accept connections" and "needs to be restarted" are completely orthogonal questions and the probes are constructed that way. "liveness" is something *kubelet* primarily acts on; "readiness" is something the *endpoints/service controller* primarily acts on

To beat the dead horse even further, I can imagine a deadlocked container still being ready--imagine only some of the threads are deadlocked--even though I'd probably want such a container to be marked unalive and restarted!

The good news is, **if those definitions don't make sense for your particular workload, you can build whatever logic you want into e.g. a shell script you tack on to the container**. I actually really strongly encourage you to do this anyway; you can make such scripts arbitrarily sophisticated. I personally do this for a toy project I run, tacking scripts onto containers I don't produce (docker makes it easy) to do in depth liveness and health checks.

To take a step back, this seems like a definitional debate, which is something I don't generally do. By "ready" Kubernetes just means "able to accept new traffic", by "live" it means "doesn't need restarting". The words "readiness" and "liveness" don't seem to mean those things to you and some other people in the thread. Perhaps we could have chosen better words to make it clearer which concepts they stood for. But we did actually intend these two concepts to be distinct and orthogonal.

**Comment by matthyx** [comment]
> I personally do this for a toy project I run, tacking scripts onto containers I don't produce (docker makes it easy) to do in depth liveness and health checks.

@lavalamp do you mind sharing it? (if possible)

**Comment by lavalamp** [comment]
> I think we should make these two concept clearer in the documentation as they are unrelated as per your comment.

Totally in favor of that.

>> I personally do this for a toy project I run, tacking scripts onto containers I don't produce (docker makes it easy) to do in depth liveness and health checks.
>
> @lavalamp do you mind sharing it? (if possible)

It's very specific, it wouldn't be a good example :)

**Comment by matthyx** [comment]
Could a native speaker tackle the documentation update? I don't mind updating the French translation once done...

**Comment by lavalamp** [comment]
Yeah that sounds reasonable. I'll see if I can find some time, but if someone is inspired don't wait for me :)

**Comment by thockin** [comment]
@chenopis might be able to help with docs.

I'll take a look at the API change.  I dislike the need for Kubelet to store state, and we'll need to be VERY VERY careful about defining the guarantees here, doubly so in the face of version skew.

**Comment by thockin** [comment]
Also to add to @lavalamp's excellent explanation: you can also add a custom HTTP server on a different port, as a container in your pod.  Pods are built EXACTLY for this!  This custom HTTP server can do anything you need - merge other health signals, execute periodic commands and cache results, check with other control-planes for advice, or even generate random numbers.

Then you can point your probe at this alternate port, which totally encapsulates all your complexity into your own space - you don't need Kubernetes to do anything more than hit your new port.

Here's an example of a healthz server that periodically runs and caches the results of a command:  https://github.com/kubernetes/contrib/tree/master/exec-healthz

**Comment by lavalamp** [comment]
> you can also add a custom HTTP server on a different port, as a container in your pod.

Yeah, that's usually better, if you can do it for your workload. One especially nice thing is that you can test multiple versions of your health checking by putting them in parallel containers. Health checking can be very tricky to test, depending on the rarity of the condition you want to catch and how easy it is to reproduce without the actual thing you want to monitor.

One case where parallel containers don't work (or at least, not easily, if you want to use the normal logging stack) is if you need to read std out/err of the process. It's super gross to have to do that, but if you have to parse logs of some random process to detect its error conditions, you have to. (My toy has to :/ )

**Comment by matthyx** [comment]
> This custom HTTP server can do anything you need - merge other health signals, execute periodic commands and cache results, check with other control-planes for advice, or even generate random numbers.

We do have something similar that is being deployed inside the same Spring context as the application, and we allow developers to write custom checks for that purpose... of course the drawback is that it takes as long as the application to be up.

I like the idea of a separate container within the same pod that will start faster, but what will happen if that health container dies:
- liveness: my application in the other container will get restarted, even if it doesn't require it
- readiness: my application will be taken out of the load balancer pool, and never put back

**Comment by lavalamp** [comment]
> ... what will happen if that health container dies ...

Configure a liveness check for the health container!

(Also, go binaries that do simple things are very fast to start and reliable. Just a thought. :) )

**Comment by matthyx** [comment]
Actually, that's exactly what I should do to use the health container at first.
Ah, right, the liveness probe will cause the container to restart, not the whole pod?

**Comment by sekhrivijay** [comment]
We want to fail fast if health is down. However,  we also want to retry a lot of times during the startup of the application just in case the app is taking long time to come up.  The problem is it uses the same configuration for startup and rest of the time the container is running. So if we retry 25 times in the beginning,  it will retry 25 times rest of its life if failure occurs while it was running . This would cause serious issue because it will not bring container out of rotation soon enough. 
 
 If we increase the initialDelay to 4 minutes with 3 retries, then even though app may come up sooner (perhaps 1 minutes) , it wont get into rotation unless 4 minutes have elapsed. The startup time of the app varies based on different factor. If there is CPU pressure , or memory pressure or app itself is doing some initial tasks,  startup time could significantly vary from time to time  , even though it is same application.
 
 One way to solve for this is to have separate configuration for startup and for rest of the time the pod is running. That way we can set a high failureThreshold for startup but low for rest of the time. Example would be 
 
```
        startupReadinessProbe:
          httpGet:
            path: /health
            port: 80
          timeoutSeconds: 1
          failureThreshold: 50
          initialDelaySeconds: 5
          periodSeconds: 3

	  

     readinessProbe:
          httpGet:
            path: /health
            port: 80
          timeoutSeconds: 1
          failureThreshold: 2
          initialDelaySeconds: DOES not apply anymore 
          periodSeconds: 5
```
The startup probe would only run  until the pods gets into rotation. Once the pod is in rotation, then regular probe would run and startup probe will stop

**Comment by matthyx** [comment]
@sekhrivijay the first idea was to wait for the readiness to succeed before checking for liveness (#70652) but then it has changed to introduce a second counter only used during startup (#71449).
There isn't much traction to it probably due to the end of the year, I will try to stir things up a little bit.

**Comment by matthyx** [comment]
Should I submit a talk for Kubecon Europe about the need of a solution for slow starting containers?

**Comment by dims** [comment]
Go for it @matthyx !

**Comment by matthyx** [comment]
Proposal submitted, wish me luck :-)

**Comment by jian-he** [comment]
> @sekhrivijay the first idea was to wait for the readiness to succeed before checking for liveness (#70652) but then it has changed to introduce a second counter only used during startup (#71449).
> There isn't much traction to it probably due to the end of the year, I will try to stir things up a little bit.

@matthyx Ran into similar issues... Could you shed some light on how the `MaxInitialFailureCount` can solve the problem of "LivenessProbe should start after ReadinessProbe Succeeded ?

**Comment by matthyx** [comment]
Sure, so the main problem with slow starting containers is that you have to leave them enough time to start before having `livenessProbe` fail `failureThreshold` times (otherwise `kubelet` kills them before they have a chance to be up).
This is even worse when load on the nodes can increase this long startup time since you have to take into account the worse case scenario (let's call it **T**).

You can handle that today with various strategies:
- set `initialDelaySeconds` > **T**: the `livenessProbe` isn't run for **T** time (so you're blind in the monitoring), you have to wait at least **T** until `kubelet` kills your container after a deadlock - this could only happen at the beginning of the container lifetime
- set `failureThreshold` * `periodSeconds` > **T**: the `livenessProbe` is run every `periodSeconds` (so you can monitor the probe status), **but you still** have to wait at least **T** until `kubelet` kills your container after a deadlock - this could happen at any time

By using `MaxInitialFailureCount` * `periodSeconds` > **T** you have the following benefits:
- `livenessProbe` is run every `periodSeconds` - good for monitoring
- container has enough time to start in the worse case scenario
- when `livenessProbe` has succeeded once, it has to fail only `failureThreshold` (should be set to 1 or at least low) before `kubelet` kills it

I hope it is more explicit this way :-)

**Comment by ejc3** [comment]
@matthyx  If you do your slow initialization work in postStart, that will delay both livenessProbe and readinessProbe from running; then those two probes only need to worry about monitoring the health of an already initialized container: https://blog.openshift.com/kubernetes-pods-life/

**Comment by jian-he** [comment]
That explains all, thanks!

**Comment by matthyx** [comment]
> @matthyx If you do your slow initialization work in postStart, that will delay both livenessProbe and readinessProbe from running; then those two probes only need to worry about monitoring the health of an already initialized container: https://blog.openshift.com/kubernetes-pods-life/

The issue is that I deal with approx 200 legacy applications that are Java based running on Tomcat or Resin. The long initialization is mostly due to some in-jar libs that take forever to start and load their configs from files or DB. I cannot justify rewriting all these apps with modern technologies, or `postStart` hooks...

**Comment by ejc3** [comment]
@matthyx you could solve that by having your preStart not return until your java app has started.

**Comment by matthyx** [comment]
> that will delay both livenessProbe and readinessProbe from running

Actually this is wrong according to the `kubelet` [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/worker.go)... and even if you look carefully at the OpenShift [link](https://blog.openshift.com/kubernetes-pods-life/) `livenessProbe` and `readinessProbe` start before `postStart` has finished...

Actually we solve this problem with a `postStart` script that runs the same command as `livenessProbe` in a **finite** loop and once that command succeeds it / or when the loop ends, stops inhibiting the regular `livenessProbe`.

On Monday I can share some actual code we're using, but this is messy and I would prefer having my [PR](https://github.com/kubernetes/kubernetes/pull/71449) merged to have cleaner code...

**Comment by matthyx** [comment]
Alright, so obviously you have to use a wrapper for `livenessProbe`:
```
#!/bin/bash
if [ -f "/var/run/post_start_done" ] ; then
    liveness_check
fi
```
Where `liveness_check` contains the actual command:
```
#!/bin/bash
httpUrl="http://localhost:80/service/health?type=LIVENESS"
curl -s -f "$httpUrl"
```
And here is the `postStart` handler also calling `liveness_check` in a finite loop:
```
#!/bin/bash
i="0"
times=${RETRY_COUNT:-"6"}
until [ "$i" -ge "$times" ] ; do
    i=$[$i+1]
    echo "running liveness_check attempt: $i"
    liveness_check
    if [ "$?" -eq "0" ] ; then
        break
    fi
    sleep 10
done
touch /var/run/post_start_done
```
As you can see, it mimics the proposed `maxInitialFailureCount` working, but is a lot more messy, requires a command (curl in the image, docker exec doing a fork, ...) and is subject to tampering `post_start_done` state file.

I would really prefer to have `kubelet` doing all that for me.

**Comment by ejc3** [comment]
Agreed, that would be cleaner and much more obvious :)

**Comment by geerlingguy** [comment]
After spending a few hours reading through this issue, all the docs, etc., I'm still finding it annoyingly complex to account for a pod where, the first time it starts up and configures everything, it could take 5-10 minutes, but then after that subsequent pods could take anywhere from < 1 second to 1 minute.

I want the pods to come up quickly, and also stop serving requests quickly if there's a failure in one of the probes, but to account for the fact that initialization the first time could take 10 minutes, I have to leave padding in both my probes (since I can't have one run before the other kicks in) and either make it so no container is 'ready' until after 10 min or so, or containers could be in a bad state (but not marked 'unhealthy') for up to 10 min or so.

I, like others earlier in this thread mentioned, would love to have an annotation or configuration option to have it be like "livenessBeforeReadiness" so I can use that kind of behavior (which seems more intuitive to me?).

**Comment by wstrange** [comment]
For the most part we have been able to solve these kinds of issues with init containers.  The init container handles the case where setup takes an unusual amount of time. For the happy path it just exits with a no-op.

**Comment by sfitts** [comment]
@wstrange as has been mentioned, for some of us, init containers aren't the solution because the software we are dealing with can't be modified in a way that would support them.  It's great when you can, but it simply isn't possible in all situations.  There are of course workaround for that, but as @matthyx has observed, it would be much better if they were unnecessary.

**Comment by matthyx** [comment]
> I have to leave padding in both my probes...

Be careful with not executing your probes, as it leaves you blind regarding the state of your container.

The main idea behind something like `maxInitialFailureCount` is to allow probes to run and fail during initialization, while still recording container states.

**Comment by geerlingguy** [comment]
@matthyx - That's exactly why I think this issue has had so much uptake. The way liveness/readiness probes currently work is quite fragile and/or nerve-wracking because you can't really optimize for the use case of "pods can take a while to start and be responsive sometimes, and that's okay".

I was looking at https://github.com/kubernetes/kubernetes/pull/71449 and it looks like there is still a decently-long path towards it hitting a stable release; therefore I have to still find ways to work around this shortcoming for probably the next year or two at least (I'm on EKS... still on 1.11 :P).

I will definitely look into using an `initContainer` for now, as that feels like it could be the cleanest way to achieve this without artificially inflating the probes' initial timeouts.

**Comment by matthyx** [comment]
> The way liveness/readiness probes currently work is quite fragile and/or nerve-wracking because you can't really optimize for the use case of "pods can take a while to start and be responsive sometimes, and that's okay".

I agree, and that's why I have submitted a talk for the next Kubecon to raise awareness and hopefully get smart minds working on an acceptable solution.

> I will definitely look into using an `initContainer` for now, as that feels like it could be the cleanest way to achieve this without artificially inflating the probes' initial timeouts.

You can also try a wrapper for the livenessProbe, as suggested in my [previous comment](https://github.com/kubernetes/kubernetes/issues/27114#issuecomment-456306347).

**Comment by tobiasvdp** [comment]
Having an init container to handle the expected behavior of a liveness probe to start checking after the readinessprobe seems counter intuitive. An implementation from the past with a configured delay before it starts checking, should not hinder the development of this expected behaviour.

**Comment by matthyx** [comment]
> Proposal submitted, wish me luck :-)

Proposal accepted, see you in Barcelona!

In the meantime, could someone merge my KEP (https://github.com/kubernetes/enhancements/pull/860)?

**Comment by from-nibly** [comment]
We ended up building a sidecar container that fronts `liveness` and `readiness`. However if the main container restarts for some reason we have to reset the `haveBeenReadyAtLeastOnce` state. We use a `postStart` hook to call the sidecar container and let it know the current container it's monitoring is new. 

This, however, is called on the first boot of the container as well. The side car is written in go so it comes up really quick but still not quick enough because it fails the first `postStart` hook with a connection refused. This ends up restarting the main container on the first boot a couple of seconds after it's started. So all of our pods now have a single restart. 

I'm wondering what the purpose of an httpGet `postStart` hook is when it's very unlikely that anything will be able to respond to it that quickly. Since it restarts the container it would never be able to come up cause it would keep restarting. Is there something I'm missing here?

Other than having an init container installing it I can't rely on a given binary like curl being present on any container we build/use. Using curl though would allow me to have retries built into the `postStart` hook. Now, everything starts getting pretty "Rube Goldberg". Then at that point why don't I just build a go binary that I side-load with an init container to directly call the main process and front all of the retry/wait until readiness logic. Again calling into question the existence of an `httpGet` `postStart` hook.

I agree with what a lot of people have already said: this feels like logic that should be abstracted into kubernetes. I don't necessarily know what the solution should be but I think the extra property in the `livenessProbe` to wait for readiness is a good start.

**Comment by mdavis-ciena** [comment]
@from-nibly - FYI, I filed [httpGet postStart hook should retry on connection refused/timeout #78286](https://github.com/kubernetes/kubernetes/issues/78286)

**Comment by matthyx** [comment]
startupProbe is alpha in 1.16!

**Comment by lostpebble** [comment]
@matthyx thanks!

From a brief look at the pull request, it looks a bit different to what was talked about here.

Is there any documentation we can look over about how this startupProbe works ?

**Comment by abdennour** [comment]
The proposed `waitForReadinessProbe` would address our use case nicely. Any update on the status of that work?

**Comment by abdennour** [comment]
@geerlingguy  we share the same case. A Laravel App took sometimes up to 4 minutes to start (php-fpm) .
Do you mind to share an example of initContainers that you implemented to overcome the issue of probes ?

**Comment by abdennour** [comment]
@geerlingguy  Looks like the new probe **startupProbe** will solve the issue here  : 🎊 🎊 🎊
 
```sh

startupProbe:
  httpGet:
    path: /healthz
    port: liveness-port
  failureThreshold: 30
  periodSeconds: 10
```

Reference : https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes

**Comment by matthyx** [comment]
Indeed!
